### ISSUE TITLE
Custom command blocks should not yield - except (maybe) when entering recursion

### ISSUE THREAD

**Issue by nate-norgaard** [jmoenig/Snap--Build-Your-Own-Blocks#1418]
![image](https://cloud.githubusercontent.com/assets/5818959/18758920/e3cf2f6a-80ae-11e6-960a-8e81f8abcf5b.png)

Hi! I'm a TEALS volunteer teacher using Snap! as our main learning tool in the classroom.

While I have been creating my own games in Snap, the bad performance of the custom "Make a block" blocks is hard to ignore. The screen capture above shows a simple test I created (the sprite was facing right, 90 degrees). These are the test results on my machine:
Script 1 - 6 seconds
Script 2 - 12 seconds
Script 3 - 6 seconds

If script 2 takes 6 additional seconds and includes about 360 calls to the custom block, the overhead for each call to a custom block is around 6 / 360 = 16.67 milliseconds. That's too high (the test shows how the script ran in double the time), especially when a loop may contain multiple custom blocks - and those custom blocks may within their bodies call other custom blocks. I don't know why the overhead is so high for such a simple custom block - I imagine there is a significant amount of room for improvement.

I want to teach my students to use custom blocks to organize and abstract, but with this performance hit, it's completely infeasible for games, and still painful for other types of Snap programs.

Here is a link to my test script:
http://snap.berkeley.edu/snapsource/snap.html#present:Username=nnorgaard&ProjectName=PerformanceTest-CustomBlocks

**Comment by dekrain** [comment]
I recommend you to use warp block.

**Comment by nate-norgaard** [comment]
Thanks for the reply dekrain,
The warp block does not solve the problem - it just skips time spent refreshing the drawing screen. In most games, we want to see the screen refresh after each loop in order to see the sprite move smoothly.

**Comment by brianharvey** [comment]
There is an option "Prefer smooth animations" in the Settings (gear) menu.  See if you like that better.

**Comment by nate-norgaard** [comment]
Thanks for the reply brianharvey,
I don't think that is the intended use of the "Prefer smooth animations" setting (I'm not sure what it's for).
When I use it without warp, it doubles the running time of all scripts. I reran 1 and 2 again.
Script 1 - 12.8
Script 2 - 26.2

When I use it with warp, we still see the whole loop executed immediately (the sprite does not visibly crawl across the screen). This would not work in a game or other application with an emphasis on visuals.

**Comment by dekrain** [comment]
Tip: use `@nickname` instead of `nickname`

**Comment by jmoenig** [comment]
Wow, you're right. And, you know what, I *think* you've discovered an actual bug in Snap's scheduler, because it's *not* the custom block overhead that's causing the slow-down, but the fact that there is a yield after every custom block. That's why the seemingly same script takes exactly twice as long when written in custom blocks than when using primitives.

This is a great catch! I believe it will help make Snap a lot faster in the next release.

Thank you!

**Comment by cycomachead** [comment]
Where in the code is that? I thought for some reason that was added intentionally, but I can't recall why.

**Comment by jmoenig** [comment]
This will be tricky, but it's clearly to be avoided. If you just create a stack of, say, 50 custom ```move1``` blocks from @nate-norgaard 's project and click on it to run it, you'll note a very smooth and gliding motion. Whereas what *should* happen is that the whole stack is evaluated all at once, just like when you stack 50 primitive ```move 1``` block together and then click an that stack.

**Comment by cycomachead** [comment]
I and others always thought that "slowness" was intentional...so if it's avoidable that will be _huge_ for everyone.

**Comment by jmoenig** [comment]
Well, it's not going to be so huge, after all, because for most of the number-crunshing projects you'll ```warp``` stuff anyway, which eliminates the yields...

**Comment by jmoenig** [comment]
Hmm... it sure looks as if I've added the yield for custom command blocks very intentially in the code... Let's see what happens if I just take it out...

**Comment by cycomachead** [comment]
Most people don't use warp, and it is the animations that cause things to be slow.

I think this is one of the main reasons people think Snap<em>!</em> is slow. Students write code like this very early on, think Snap<em>!</em> is slow, and then never change their minds.

**Comment by jmoenig** [comment]
... okay, this looks too easy. If I just take the ```yield``` the performance difference @nate-norgaard notices simply goes away, and both block versions are executed exactly the same. Now, if only I could remember *why* I put it in. ...

**Comment by cycomachead** [comment]
What line / file is this one at? I remember finding it before, but couldn't quickly remember from skimming the code.

**Comment by jmoenig** [comment]
Ha! Now I remember again. The reason for yielding is recursion! When drawing a fractal it is often nice to see the steps in which it is drawn. Also, when you use recursion for your custom control structures like ```while``` or ```for each``` you probably want to yield at the end of every loop body, or when entering or leaving a recursive call.

So, that's the issue here. Looks like we either need an explicit ```yield``` block (which I don't like), or detect recursion inside the evaluator (which is hard, because we can't compare blocks for equality).

Maybe we should actually add a ```yield``` block. We could name it ```wait``` (without any input). Or we could somehow make the existing ```wait``` block variadic, like in GP. That way programmers could decide for themselves when to wait for a screen refresh when drawing a fractal.

**Comment by jmoenig** [comment]
... it's at the end of ```Process.prototype.evaluateCustomBlock```

**Comment by jmoenig** [comment]
... of course, changing this would affect zillions of existing projects. They would all need to manually insert ```wait```s so they could run the same as before...

But it could totally be worth it.

**Comment by cycomachead** [comment]
D'oh! I totally skipped over that.... 

If they run "fine" -- i.e. The output is the same, then I think it's worth the improvement. 

In terms of recursion -- the yielding is only helpful for fractals/visual, not list operations and such, right? In those cases, non-warped recursion would see a benefit? I'm not even 100% sure the drawing of fractals is always desired. Certainly for learning, but we also have single stepping for that! :D 

In that case, I really like variadic `wait`. 

It's always easier for a user to slow Snap! down themselves, than to speed it up.

**Comment by nate-norgaard** [comment]
Wow, thanks @jmoenig and @cycomachead for looking into this!
If the extra time was truly intended to slow down the blocks, then please take it out. I agree with @cycomachead that slowing Snap down is much easier than speeding it up.

Thanks for your work!

**Comment by nate-norgaard** [comment]
Also, @jmoenig, it seems as though you're hesitant to remove the yield for recursive cases. I still think all yield should be removed because of the perf limitation.

However, you _could_ add one of the checkbox settings to Snap! along the lines of "Increase yield for recursion", "Slow down recursion", or even "Slow down custom blocks" that is unchecked by default but available as a teaching opportunity.

**Comment by cycomachead** [comment]
Here's a case I think we might have discussed a few years ago:

Infinite Recursion? If we remove the yield, can you still stop things manually?
I remember at one point years ago this was tricky.

**Comment by brianharvey** [comment]
Radical proposal: When visual stepping is enabled and the slider is all the way 
to the right, it doesn't actually light up blocks, but it /does/ yield when 
calling custom blocks. This slightly makes sense because there /is/ a visual 
indication of the stepping, just on the stage instead of in the scripts.

**Comment by brianharvey** [comment]
Speaking of which, does turbo mode turn off that extra yield? Maybe that's 
enough of a mechanism.

**Comment by Jonathan50** [comment]
>So, that's the issue here. Looks like we either need an explicit yield block (which I don't like), or detect recursion inside the evaluator (which is hard, because we can't compare blocks for equality).
>
Maybe we should actually add a yield block. We could name it wait (without any input). Or we could somehow make the existing wait block variadic, like in GP. That way programmers could decide for themselves when to wait for a screen refresh when drawing a fractal.

I think this works:
![yield](https://cloud.githubusercontent.com/assets/13597222/18770984/19e24cbe-818f-11e6-9c98-cbdddfcc146d.png)

**Comment by jmoenig** [comment]
Wow, lots of misunderstandings in this thread about what "slow down" means in a multiplexed environment that time-shares processes, including the UI. Also, the difference between ```turbo``` and ```warp``` is apparently not clear.

Anyway: @nate-norgaard , would you please do some quick tests with my current in-development version? I've made some changes that - hopefully - let us have our pie and eat it, too. Here's a link to your project opening in the dev version:

http://snap.berkeley.edu/snapsource/dev/snap.html#cloud:Username=nnorgaard&ProjectName=PerformanceTest-CustomBlocks

You should be able to see that all your test scripts now take about the same time to execute, roughly 6 seconds. The idea is that custom blocks should behave exactly the same as the built-in primitives, because obviously otherwise the whole idea of procedures doesn't advertise very well. It's great that you're encouraging your students to write custom blocks for many little abstractions, and there should not be any performance penalty involved in that. I'm confident that my changes to the evaluator guarantee this maxim now.

HOWEVER, I also want kids to explore recursion in a safe and encouraging environment, and that means that they have to be able to observe what's going on and to stop it, if it runs forever. Therefore, I want kids to try things like this script (inspired by *Mindstorms*):

![spin](https://cloud.githubusercontent.com/assets/2596285/18787081/86ed4022-81a1-11e6-9ad2-9b865a017d62.png)

and actually see the sprite slowly turn forever. Since this example (like the one in *Mindstorms*) doesn't have a base case it also needs to be stoppable, and for this it needs to yield control to the IDE. In such infinite recursion cases Snap goes into "emergency-yielding" mode which slows down the whole IDE to a "heartbeat" of 2 Hz, just barely enough to let you hit the red stop button. Clearly a fun and interactive learning experience doesn't happen in "emergency" mode. That's why I have implemented a check for direct recursion for each custom block. If a custom command block definition is directly recursive it will still yield once, unless it is ```warped```. I believe that this is the best of both worlds, because you *can* make recursive procedures "fast" simply by warping them. However, by default you can see what's going on and interact with that.

BTW, all of this only applies to commands, reporters are *always* unyielding (```warped```).

Another example where I think automatic yielding makes sense is control structures such as ```while```:

![while](https://cloud.githubusercontent.com/assets/2596285/18787663/12e0b79c-81a4-11e6-9cff-18cafbd0cf0b.png)

This is so easy to do, and it should be fun! Yet, if we don't yield whenever we call the definition for ```while``` the little test script will simply run for a while with nothing visible happening in the stage, and then all of the sudden place the sprite right next to where it touches the edge. That's why I'm convinced that auto-yielding in cases of direct recursion is the right thing to do. Again, you can totally "speed" up this definition of ```while``` (and of all recursive blocks) by wrappig a ```warp``` block around the script with the recursive call:

![warped while](https://cloud.githubusercontent.com/assets/2596285/18787850/dba814a4-81a4-11e6-972f-329ba65e89f8.png)

But the point is that once you're ready to "speed up" this operation you already know what you're doing.

So, now you're asking about *indirect* recursion, right? Because those will *no longer* automatically yield in the new version. What If you *want* them to yield for smooth animations? For this purpose I've introduced a special case for ```wait```: If you insert a ```wait 0 secs``` block (or just leave the input slot empty), Snap will yield *just once* (without checking elapsed time and interpolating the execution), but only if it is not currently ```warped```. This means that you can insert ```wait 0 secs``` into your scripts whereever you feel the need for updating the display or letting the user interact, and you can *still* override those yields with ```warp```.

If you find the time, please give the dev version a try and let me know whether any of this makes sense.

Thanks!

**Comment by nate-norgaard** [comment]
Thank you @jmoenig for the patient explanation and hard work!
Since the beginning of the thread, my understanding of the yield has changed a lot, and I think understand the predicament now. I opened the test scripts from the link you provided, and they do all run at the same speed! Amazing! I'm astounded by the improvement. I'll be trying out some more speed tests on the dev environment when I get the chance. Do you know when these changes will go live?

Thanks for the prompt help @jmoenig and everyone else involved!

**Comment by cycomachead** [comment]
> BTW, all of this only applies to commands, reporters are always unyielding (warped).

Does this mean that the `wrap` in many of the tools blocks is unnecessary?

**Comment by cycomachead** [comment]
So I tried the `spin` example in the new version, and it seems _very_ slow, slower than the normal version. 

I don't know why, I refreshed and tried again a the appearance seemed more comparable the second time.

However, I'm not at all convinced this yielding is necessarily right... The spin example takes seconds to make a full circle, where as the `warped` version is very quick to go around, but doesn't animate well. I'm not sure that either of these is more or less pedagogically correct. Our target audience (teenagers and above) will have spent their entire lives at this point with very fast computers. (We're not too many years away from a point where high schoolers will have been born after the iPhone was launched...which is weird!) The point is that, the idea that something is happening faster than they can see may not be all that uncommon. When kids have spent time interacting with digital devices, I think their expectations change on fundamental levels.

The one problem with always yielding is that in cases where computation slows things down (like rotating a sprite while it is `say`ing something is very very slow), the yields seem to do more harm than good. 

I'm not advocating for never auto-yielding, and I _definitely_ think these changes are an improvement -- I just am not sure about how often and when.

**Comment by liam4** [comment]
@cycomachead [`wrap`?](https://github.com/jmoenig/Snap--Build-Your-Own-Blocks/issues/1418#issuecomment-249256655) :P

**Comment by jmoenig** [comment]
RE:
> Does this mean that the warp in many of the tools blocks is unnecessary?

Yes. Mostly. Although back then when we first wrote them ```warp``` did help, but we've changed Snap's execution since then (about 2 years ago, I think)

RE:
> So I tried the spin example in the new version, and it seems very slow, slower than the normal version.

Hmm... I wrote a little test script and measured the time it takes to complete a full turn, and it was the exact some int he current stable version and in the dev one.

See, again "slowness" is influenced by many factors, the most obvious one being the amount to turn at each step.

RE:
> P.S. In the while example: Isn't the second call test wrong, shouldn't it be test?

No, that one's right, because I declared the ```test``` input to be of type ```Boolean unevaluated```, not a plain ring. 

RE:
> Do you know when these changes will go live?

That'll take a couple of more weeks, perhaps even months. There are some more features and fixes we need to get into the next release, and we certainly need to do more regression testing on this particular change. Plus, we want to be really careful with new updates while classes are in session at BJC and TEALS schools, especially since is the first batch of the new APCSP courses.
